File: 3143. Maximum Points Inside the Square.go
func maxPointsInsideSquare(points [][]int, s string) int {
	pList := make([]*Point, len(points))
	for i := range(points) {
		pList[i] = &Point{x: points[i][0], y: points[i][1], tag: s[i]}
	}
	sort.Slice(pList, func(i, j int) bool {
		return pList[i].distance() < pList[j].distance()
	})
	set := make(map[byte]bool)
	answer := 0
	i := 0
	for i < len(pList) {
		distance := pList[i].distance()
		subset := make(map[byte]bool)
		for i < len(pList) && pList[i].distance() == distance {
			if _, ok := set[pList[i].tag]; ok {
				return answer
			}
			if _, ok := subset[pList[i].tag]; ok {
				return answer
			}
			subset[pList[i].tag] = true
			i += 1
		}
		answer = i
		for key := range subset {
			set[key] = true
		}	
	}
	return answer
}

type Point struct {
	x, y int
	tag byte
}

func (p *Point) distance() int {
	return max(abs(p.x), abs(p.y))
}

func abs(i int) int {
	if i < 0 { return -i }
	return i
}

File: 3143. Maximum Points Inside the Square.py
class Solution:
    def maxPointsInsideSquare(self, points: List[List[int]], s: str) -> int:
        p_list = []
        for i in range(len(points)):
            p_list.append(Point(points[i][0], points[i][1], s[i]))
        p_list.sort(key=lambda x: x.distance())
        set_p = set()
        answer = 0
        i = 0
        while i < len(p_list):
            distance = p_list[i].distance()
            subset = set()
            while i < len(p_list) and p_list[i].distance() == distance:
                if p_list[i].tag in set_p or p_list[i].tag in subset:
                    return answer
                subset.add(p_list[i].tag)
                i += 1
            answer = i
            set_p |= subset
        return answer

class Point:
    def __init__(self, x, y, tag):
        self.x = x
        self.y = y
        self.tag = tag
    def distance(self):
        return max(abs(self.x), abs(self.y))
